services/repository/src/main/java/org/sagebionetworks/auth/AuthenticationFilter.java

```diff
@@ -46,126 +43,111 @@
+				try {
+					oidcTokenHelper.validateJWT(bearerToken);
+				} catch (IllegalArgumentException e) {
+					String failureReason = "Invalid access token";
+					if (StringUtils.isNotEmpty(e.getMessage())) {
+						failureReason = e.getMessage();
+					}
+					HttpAuthUtil.reject((HttpServletResponse)servletResponse, failureReason);
+					log.warn(failureReason, e);
+					return;
+				}	
+			}
 		}
-		
-		if (userId == null && !allowAnonymous) {
-			String reason = "The session token provided was missing, invalid or expired.";
-			reject(req, (HttpServletResponse) servletResponse, reason);
-			log.warn("Anonymous not allowed");
-			return;
-		}
-		
+
 		// If the user has been identified, check if they have accepted the terms of use
 		if (userId != null) {
-			boolean toUCheck = false;
-			try {
-				toUCheck = authenticationService.hasUserAcceptedTermsOfUse(userId);
-			} catch (NotFoundException e) {
-				String reason = "User " + userId + " does not exist";
-				reject(req, (HttpServletResponse) servletResponse, reason, HttpStatus.NOT_FOUND);
-				log.error("This should be unreachable", e);
+			if (!authenticationService.hasUserAcceptedTermsOfUse(userId)) {
+				HttpAuthUtil.reject((HttpServletResponse) servletResponse, TOU_UNSIGNED_REASON, HttpStatus.FORBIDDEN);
 				return;
-			}
-			if (!toUCheck) {
-				String reason = "Terms of use have not been signed";
-				reject(req, (HttpServletResponse) servletResponse, reason, HttpStatus.FORBIDDEN);
+			}	
+		} else if (bearerToken!=null) {
+			try {
+				if (!authenticationService.hasUserAcceptedTermsOfUse(bearerToken)) {
+					HttpAuthUtil.reject((HttpServletResponse) servletResponse, TOU_UNSIGNED_REASON, HttpStatus.FORBIDDEN);
+					return;
+				}
+			} catch (Exception e) {
+				HttpAuthUtil.reject((HttpServletResponse) servletResponse, e.getMessage(), HttpStatus.FORBIDDEN);
 				return;
 			}	
-		}
-		
-		if (userId == null) {
-			userId = BOOTSTRAP_PRINCIPAL.ANONYMOUS_USER.getPrincipalId();
+		} else { // anonymous
+				userId = BOOTSTRAP_PRINCIPAL.ANONYMOUS_USER.getPrincipalId();
+				// there is no bearer token (as an Auth or sessionToken header) or digital signature
+				// so create an 'anonymous' bearer token
+				bearerToken = oidcTokenHelper.createAnonymousAccessToken();
 		}
 
 		// Put the userId on thread local, so this thread always knows who is calling
 		currentUserIdThreadLocal.set(userId);
 		try {
 			// Pass along, including the user ID
 			Map<String, String[]> modParams = new HashMap<String, String[]>(req.getParameterMap());
-			modParams.put(AuthorizationConstants.USER_ID_PARAM, new String[] { userId.toString() });
-			HttpServletRequest modRqst = new ModHttpServletRequest(req, null, modParams);
+			if (userId!=null) {
+				modParams.put(AuthorizationConstants.USER_ID_PARAM, new String[] { userId.toString() });
+			}
+			Map<String, String[]> modHeaders = HttpAuthUtil.filterAuthorizationHeaders(req);
+			HttpAuthUtil.setBearerTokenHeader(modHeaders, bearerToken);
```

Comment:
set not as Authorization header but as a Synapse-specific header name so as not to collide with Basic auth header used in special purpose filters